### DISCUSSION TITLE
Combine soft Time windows with hard Time Windows in pickups and deliveries

### DISCUSSION THREAD

**Discussion by jmonkiwi** [google/or-tools#2492] [Routing (and legacy CP) questions] [answered]
Hi everyone,

I'm currently working in a pickup and deliveries problem with time windows, and I want to have hard time windows constraints for the pickups, so the vehicle doesn't visit the pickup node before the store opens, and soft time windows for the delivery nodes. 

I'm using this code to model this, but it seems to be not working, as the vehicles are visiting the pickups node before the time windows defined for the pickups.

```python
for location_idx, time_window in enumerate(time_windows):
        if location_idx in [data['depot']]:
            continue
        elif location_idx in deliveries:
            index = manager.NodeToIndex(location_idx)
            print(index)
            time_dimension.SetCumulVarSoftLowerBound(index, time_windows[index][0], 10000000)
            time_dimension.SetCumulVarSoftUpperBound(index, time_windows[index][1], 10000000)
        elif location_idx in pickups:
            index = manager.NodeToIndex(location_idx)
            time_dimension.CumulVar(index).SetRange(time_windows[index][0], time_windows[index][1])
```

Any suggestions of how to model this? I would appreciate any help.
Thank you very much.

**Comment by Mizux** [comment, marked as answer]
simply replace `time_windows[index]` by `time_window` (ed prefered) **OR** `time_windows[location_idx]`.

i.e. don't use `index` for your own time_windows object...

**Reply by jmonkiwi** [reply]
Hi Mizux, thanks for the help. 
I try it and it stills the same, the time window for the store is violated, is there anything else I could try?

**Reply by Mizux** [reply]
1) Please don't use `time_windows[index]`, here `index` is a solver internal index while your `time_windows` must be indexed using your own indexing order. i.e. methods manager.NodeToIndex() and manager.IndexToNode() are used to do the conversion between both representation.

2) Could you provide a full example so we can reproduce the error and help you. I'm pretty confident it should just work...

**Reply by jmonkiwi** [reply]
I already found the problem, I put something wrong, now it works!! Thank you very much for the help!!
